### ISSUE TITLE
Segfault on `join()` when using any traits with conflict resolution

### ISSUE THREAD

**Issue by coderstephen** [krakjoe/pthreads#475]
I'm experiencing an extremely odd error when using traits in the same program as pthreads. If two traits define the same method name, and a class uses both of those traits, and resolves the method name conflict with `insteadof`, creating and `join()`-ing a thread later in the same program causes a segmentation fault. Here's an example to explain the issue:

```php
<?php
trait TraitA
{
    public function foo()
    {
    }
}

trait TraitB
{
    public function foo()
    {
    }
}

class Child
{
    use TraitA, TraitB {
        TraitA::foo insteadof TraitB;
    }
}

class TestThread extends Thread
{
    public function run()
    {
        sleep(1);
        print "Hello from thread!\n";
    }
}

$thread = new TestThread();
$thread->start(PTHREADS_INHERIT_ALL);

// Join here causes a segfault just before finishing.
$thread->join();
print "Done!\n";
```

The output is:

    Hello from thread!
    Segmentation fault

The `sleep(1)` inside the `TestThread::run()`, when run, reveals that `join()` successfully waits for the thread to finish, but then segfaults during the cleanup process. Using Valgrind gives the following stack trace for the segfault:

```
==2582== Thread 2:
==2582== Invalid read of size 8
==2582==    at 0x866B67: _zend_mm_free_int (zend_alloc.c:2075)
==2582==    by 0x8881D4: _destroy_zend_class_traits_info (zend_opcode.c:262)
==2582==    by 0x888377: destroy_zend_class (zend_opcode.c:314)
==2582==    by 0x8A1E4E: i_zend_hash_bucket_delete (zend_hash.c:182)
==2582==    by 0x8A1E4E: zend_hash_bucket_delete (zend_hash.c:192)
==2582==    by 0x8A3B94: zend_hash_reverse_apply (zend_hash.c:733)
==2582==    by 0x883467: shutdown_executor (zend_execute_API.c:300)
==2582==    by 0x894A3D: zend_deactivate (zend.c:960)
==2582==    by 0x8251A7: php_request_shutdown (main.c:1883)
==2582==    by 0x1044CC4C: pthreads_routine (object.c:1081)
==2582==    by 0xCBAB0A3: start_thread (pthread_create.c:309)
==2582==    by 0xCEA604C: clone (clone.S:111)
==2582==  Address 0x11 is not stack'd, malloc'd or (recently) free'd
==2582== 
==2582== 
==2582== Process terminating with default action of signal 11 (SIGSEGV)
==2582==  Access not within mapped region at address 0x11
==2582==    at 0x866B67: _zend_mm_free_int (zend_alloc.c:2075)
==2582==    by 0x8881D4: _destroy_zend_class_traits_info (zend_opcode.c:262)
==2582==    by 0x888377: destroy_zend_class (zend_opcode.c:314)
==2582==    by 0x8A1E4E: i_zend_hash_bucket_delete (zend_hash.c:182)
==2582==    by 0x8A1E4E: zend_hash_bucket_delete (zend_hash.c:192)
==2582==    by 0x8A3B94: zend_hash_reverse_apply (zend_hash.c:733)
==2582==    by 0x883467: shutdown_executor (zend_execute_API.c:300)
==2582==    by 0x894A3D: zend_deactivate (zend.c:960)
==2582==    by 0x8251A7: php_request_shutdown (main.c:1883)
==2582==    by 0x1044CC4C: pthreads_routine (object.c:1081)
==2582==    by 0xCBAB0A3: start_thread (pthread_create.c:309)
==2582==    by 0xCEA604C: clone (clone.S:111)
```

This problem does not occur when not passing `PTHREADS_INHERIT_CLASSES` to `Thread::start()`. It also does not occur if the classes are not defined until after the call to start, since the traits are not being inherited:

```php
<?php
class TestThread extends Thread
{
    public function run()
    {
        sleep(1);
        print "Hello from thread!\n";
    }
}

$thread = new TestThread();
$thread->start(PTHREADS_INHERIT_ALL);

trait TraitA
{
    public function foo()
    {
    }
}

trait TraitB
{
    public function foo()
    {
    }
}

class Child
{
    use TraitA, TraitB {
        TraitA::foo insteadof TraitB;
    }
}

$thread->join();
print "Done!\n";

/* Outputs:
        Hello from thread!
        Done!
*/
```

Relevant environment information:

```
$ php -i | head -n 24
phpinfo()
PHP Version => 5.6.10-dev

System => Linux php7dev 3.2.0-4-amd64 #1 SMP Debian 3.2.68-1+deb7u1 x86_64
Build Date => May 24 2015 22:21:04
Configure Command =>  './configure'  '--with-apxs2=/usr/bin/apxs2' '--enable-zend-signals' '--with-gd' '--without-pear' '--with-jpeg-dir=/usr' '--with-png-dir=/usr' '--with-vpx-dir=/usr' '--with-freetype-dir=/usr' '--with-t1lib=/usr' '--enable-gd-native-ttf' '--enable-exif' '--with-config-file-path=/etc/php56' '--with-config-file-scan-dir=/etc/php56/conf.d' '--with-mysql=/usr' '--with-mysql-sock=/var/run/mysqld/mysqld.sock' '--with-zlib' '--with-gmp' '--with-zlib-dir=/usr' '--with-gettext' '--with-kerberos' '--with-imap-ssl' '--with-mcrypt=/usr/local' '--with-iconv' '--enable-sockets' '--with-openssl' '--with-pspell' '--with-pdo-mysql=mysqlnd' '--with-pdo-sqlite' '--with-pgsql' '--with-pdo-pgsql' '--enable-soap' '--enable-zip' '--enable-xmlreader' '--enable-phar=shared' '--with-xsl' '--enable-ftp' '--enable-cgi' '--with-curl=/usr' '--with-tidy' '--with-xmlrpc' '--enable-mbstring' '--enable-sysvsem' '--enable-sysvshm' '--enable-shmop' '--with-readline' '--enable-pcntl' '--enable-fpm' '--enable-intl' '--with-imap' '--with-mysqli=mysqlnd' '--enable-calendar' '--disable-debug' '--enable-maintainer-zts' '--prefix=/usr/local/php56-zts'
Server API => Command Line Interface
Virtual Directory Support => enabled
Configuration File (php.ini) Path => /etc/php56
Loaded Configuration File => /etc/php56/php-cli.ini
Scan this dir for additional .ini files => /etc/php56/conf.d
Additional .ini files parsed => (none)
PHP API => 20131106
PHP Extension => 20131226
Zend Extension => 220131226
Zend Extension Build => API220131226,TS
PHP Extension Build => API20131226,TS
Debug Build => no
Thread Safety => enabled
Zend Signal Handling => enabled
Zend Memory Manager => enabled
Zend Multibyte Support => provided by mbstring
IPv6 Support => enabled
DTrace Support => disabled
```

```
$ php -i | grep -A 2 pthreads
pthreads

Version => 2.0.10
```

**Comment by krakjoe** [comment]
Can you try that ?

**Comment by coderstephen** [comment]
Excellent -- that definitely fixed the issue for me on PHP 5.6.10, with all uses of traits that I can come up with. The output for the first script now gives the correct output:

```
Hello!
Done!
```

It also now works properly on PHP 5.4.42 and 5.5.26.

Thanks!

**Comment by krakjoe** [comment]
Excellent, thanks for feedback ;)
